rspirv/sr/context.rs

```diff
@@ -12,11 +12,89 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+use std::{
+    collections::HashMap,
+    hash::BuildHasherDefault,
+    marker::PhantomData,
+};
+
+use fxhash::FxHasher;
 use spirv;
 
-use super::{Type, TypeToken, Constant, ConstantToken};
-use sr::constants::ConstantEnum;
-use sr::types::{StructMember, TypeEnum};
+use sr::constants::{Constant, ConstantEnum};
+use sr::types::{StructMember, TypeEnum, Type};
+
+type FastHashMap<K, V> = HashMap<K, V, BuildHasherDefault<FxHasher>>;
+type Index = u32;
+
+#[derive(Debug)]
+pub struct Token<T> {
+    index: Index,
+    marker: PhantomData<T>,
+}
+
+impl<T> Clone for Token<T> {
+    fn clone(&self) -> Self {
+        Token {
+            index: self.index,
+            marker: self.marker,
+        }
+    }
+}
+impl<T> Copy for Token<T> {}
+impl<T> PartialEq for Token<T> {
+    fn eq(&self, other: &Self) -> bool {
+        self.index == other.index
+    }
+}
+impl<T> Eq for Token<T> {}
+
+impl<T> Token<T> {
+    pub(in sr) fn new(index: Index) -> Self {
+        Token {
+            index,
+            marker: PhantomData,
+        }
+    }
+}
+
+/// A structure holding some kind of a structured SPIRV data that can be
```

Comment:
... some kind of SPIR-V entity (e.g., type, constant, instruction, etc.) that can be referenced.